File: java/prime-factors/src/main/java/PrimeFactors.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class PrimeFactors {

    private static final long FIRST_DIVISOR = 2;

    public static List<Long> getForNumber(final long naturalNumber) {
        return computePrimeFactors(Collections.emptyList(), naturalNumber, FIRST_DIVISOR);
    }

    private static List<Long> computePrimeFactors(
            final List<Long> primeFactors,
            final long naturalNumberRemainder,
            final long divisor
    ) {
        if (isRemainderOne(naturalNumberRemainder))
            return Collections.unmodifiableList(primeFactors);

        else if (isRemainderDivisibleByDivisor(naturalNumberRemainder, divisor))
            return Collections.unmodifiableList(computePrimeFactors(
                    new ArrayList<Long>(primeFactors) {{ add(divisor); }},
                    naturalNumberRemainder / divisor,
                    divisor
            ));

        else
            return Collections.unmodifiableList(computePrimeFactors(primeFactors, naturalNumberRemainder, divisor + 1));
    }

    private static boolean isRemainderOne(final long remainingNaturalNumber) {
        return remainingNaturalNumber == 1;
    }

    private static boolean isRemainderDivisibleByDivisor(final long remainder, final long divisor) {
        return remainder % divisor == 0;
    }
}

File: ecmascript/prime-factors/prime-factors.js
const FIRST_DIVISOR = 2;

export default class PrimeFactors {
  for(naturalNumber) {
    return computePrimeFactors([], naturalNumber, FIRST_DIVISOR);
  }
}

const computePrimeFactors = (primeFactors, naturalNumberRemainder, divisor) => {
  if (isRemainderOne(naturalNumberRemainder))
    return primeFactors;

  else if (isRemainderDivisible(naturalNumberRemainder, divisor))
    return computePrimeFactors([...primeFactors, divisor], naturalNumberRemainder / divisor, divisor);

  else
    return computePrimeFactors(primeFactors, naturalNumberRemainder, divisor + 1);
};

const isRemainderOne = (remainingNaturalNumber) => remainingNaturalNumber === 1;

const isRemainderDivisible = (remainder, divisor) => remainder % divisor === 0;
